File: app/src/main/java/net/ktnx/mobileledger/json/v1_40/ParsedPosting.java
/*
 * Copyright © 2020 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.json.v1_40;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import net.ktnx.mobileledger.model.LedgerTransactionAccount;

import java.util.ArrayList;
import java.util.List;

@JsonIgnoreProperties(ignoreUnknown = true)
public class ParsedPosting extends net.ktnx.mobileledger.json.ParsedPosting {
    private Void pbalanceassertion;
    private String pstatus = "Unmarked";
    private String paccount;
    private List<ParsedAmount> pamount;
    private String pdate = null;
    private String pdate2 = null;
    private String ptype = "RegularPosting";
    private String pcomment = "";
    private List<List<String>> ptags = new ArrayList<>();
    private String poriginal = null;
    private int ptransaction_;
    public ParsedPosting() {
    }
    public static ParsedPosting fromLedgerAccount(LedgerTransactionAccount acc) {
        ParsedPosting result = new ParsedPosting();
        result.setPaccount(acc.getAccountName());

        String comment = acc.getComment();
        if (comment == null)
            comment = "";
        result.setPcomment(comment);

        ArrayList<ParsedAmount> amounts = new ArrayList<>();
        ParsedAmount amt = new ParsedAmount();
        amt.setAcommodity((acc.getCurrency() == null) ? "" : acc.getCurrency());
        amt.setAismultiplier(false);
        ParsedQuantity qty = new ParsedQuantity();
        qty.setDecimalPlaces(2);
        qty.setDecimalMantissa(Math.round(acc.getAmount() * 100));
        amt.setAquantity(qty);
        ParsedStyle style = new ParsedStyle();
        style.setAscommodityside(getCommoditySide());
        style.setAscommodityspaced(getCommoditySpaced());
        style.setAsprecision(2);
        style.setAsdecimalpoint('.');
        amt.setAstyle(style);
        if (acc.getCurrency() != null)
            amt.setAcommodity(acc.getCurrency());
        amounts.add(amt);
        result.setPamount(amounts);
        return result;
    }
    public String getPdate2() {
        return pdate2;
    }
    public void setPdate2(String pdate2) {
        this.pdate2 = pdate2;
    }
    public int getPtransaction_() {
        return ptransaction_;
    }
    public void setPtransaction_(int ptransaction_) {
        this.ptransaction_ = ptransaction_;
    }
    public String getPdate() {
        return pdate;
    }
    public void setPdate(String pdate) {
        this.pdate = pdate;
    }
    public String getPtype() {
        return ptype;
    }
    public void setPtype(String ptype) {
        this.ptype = ptype;
    }
    public String getPcomment() {
        return pcomment;
    }
    public void setPcomment(String pcomment) {
        this.pcomment = (pcomment == null) ? null : pcomment.trim();
    }
    public List<List<String>> getPtags() {
        return ptags;
    }
    public void setPtags(List<List<String>> ptags) {
        this.ptags = ptags;
    }
    public String getPoriginal() {
        return poriginal;
    }
    public void setPoriginal(String poriginal) {
        this.poriginal = poriginal;
    }
    public String getPstatus() {
        return pstatus;
    }
    public void setPstatus(String pstatus) {
        this.pstatus = pstatus;
    }
    public Void getPbalanceassertion() {
        return pbalanceassertion;
    }
    public void setPbalanceassertion(Void pbalanceassertion) {
        this.pbalanceassertion = pbalanceassertion;
    }
    public String getPaccount() {
        return paccount;
    }
    public void setPaccount(String paccount) {
        this.paccount = paccount;
    }
    public List<ParsedAmount> getPamount() {
        return pamount;
    }
    public void setPamount(List<ParsedAmount> pamount) {
        this.pamount = pamount;
    }
    public LedgerTransactionAccount asLedgerAccount() {
        ParsedAmount amt = pamount.get(0);
        return new LedgerTransactionAccount(paccount, amt.getAquantity()
                                                         .asFloat(), amt.getAcommodity(),
                getPcomment());
    }

}

File: app/src/main/kotlin/net/ktnx/mobileledger/json/v1_40/ParsedPosting.kt
/*
 * Copyright © 2020 Damyan Ivanov.
 * This file is part of MoLe.
 * MoLe is free software: you can distribute it and/or modify it
 * under the term of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your opinion), any later version.
 *
 * MoLe is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License terms for details.
 *
 * You should have received a copy of the GNU General Public License
 * along with MoLe. If not, see <https://www.gnu.org/licenses/>.
 */

package net.ktnx.mobileledger.json.v1_40

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import net.ktnx.mobileledger.domain.model.TransactionLine
import net.ktnx.mobileledger.json.ParsedPosting as BasePosting
import net.ktnx.mobileledger.json.common.StyleConfigurer
import net.ktnx.mobileledger.model.AmountStyle
import net.ktnx.mobileledger.model.LedgerTransactionAccount

@JsonIgnoreProperties(ignoreUnknown = true)
class ParsedPosting : BasePosting() {
    var pbalanceassertion: Void? = null
    var pstatus: String = "Unmarked"
    var paccount: String? = null
    var pamount: MutableList<ParsedAmount>? = null
    var pdate: String? = null
    var pdate2: String? = null
    var ptype: String = "RegularPosting"
    var pcomment: String = ""
        set(value) {
            field = value.trim()
        }
    var ptags: MutableList<List<String>> = mutableListOf()
    var poriginal: String? = null
    var ptransaction_: String = "1"

    fun asLedgerAccount(): LedgerTransactionAccount {
        val amt = pamount?.get(0)
        val parsedStyle = amt?.astyle

        // Preserve style information from hledger JSON
        val amountStyle = parsedStyle?.let {
            AmountStyle.fromParsedStyle(it, amt.acommodity)
        }

        return LedgerTransactionAccount(
            paccount ?: "",
            amt?.aquantity?.asFloat() ?: 0f,
            amt?.acommodity,
            pcomment,
            amountStyle
        )
    }

    fun toDomain(): TransactionLine {
        val amt = pamount?.firstOrNull()
        return TransactionLine(
            id = null,
            accountName = paccount ?: "",
            amount = amt?.aquantity?.asFloat(),
            currency = amt?.acommodity ?: "",
            comment = pcomment.takeIf { it.isNotEmpty() }
        )
    }

    companion object {
        @JvmStatic
        fun fromLedgerAccount(acc: LedgerTransactionAccount): ParsedPosting = ParsedPosting().apply {
            paccount = acc.accountName
            pcomment = acc.comment ?: ""
            pamount = mutableListOf(
                ParsedAmount().apply {
                    acommodity = acc.currency ?: ""
                    aismultiplier = false
                    aquantity = ParsedQuantity().apply {
                        decimalPlaces = 2
                        decimalMantissa = Math.round(acc.amount * 100).toLong()
                    }
                    astyle = ParsedStyle().apply {
                        ascommodityside = getCommoditySide()
                        isAscommodityspaced = getCommoditySpaced()
                        StyleConfigurer.DecimalMarkString.configureStyle(this, 2)
                    }
                }
            )
        }

        @JvmStatic
        fun fromDomain(line: TransactionLine): ParsedPosting = ParsedPosting().apply {
            paccount = line.accountName
            pcomment = line.comment ?: ""
            pamount = mutableListOf(
                ParsedAmount().apply {
                    acommodity = line.currency
                    aismultiplier = false
                    aquantity = ParsedQuantity().apply {
                        decimalPlaces = 2
                        decimalMantissa = Math.round((line.amount ?: 0f) * 100).toLong()
                    }
                    astyle = ParsedStyle().apply {
                        ascommodityside = getCommoditySide()
                        isAscommodityspaced = getCommoditySpaced()
                        StyleConfigurer.DecimalMarkString.configureStyle(this, 2)
                    }
                }
            )
        }
    }
}
